lib/auth/auth_with_roles.go

```diff
@@ -1742,6 +1742,16 @@ func (a *ServerWithRoles) GetUser(name string, withSecrets bool) (types.User, er
 	return a.authServer.Identity.GetUser(name, withSecrets)
 }
 
+// GetCurrentUser returns current user as seen by the server.
+// Useful especially in the context of remote clusters which perform role and trait mapping.
+func (a *ServerWithRoles) GetCurrentUser(ctx context.Context) (types.User, error) {
```

Comment:
Without any additional RBAC checks here, I'm a little worried about privilege escalation since we're potentially letting the root's users expand the scope of what they can learn about the leaf cluster. For example, role names - currently they are not available anywhere in the root so users can only see them if they have proper permissions.

I think we should require at least permissions to either read the user's roles or list roles prior to returning this. If the leaf doesn't allow root users view the roles, they're not also gonna be able to see "allowed database users" - which makes sense.

Any other fields are a part of User object we may want to protect?

Reply:
As long as the user has the role, they are allowed to see its definition ([from here](https://github.com/gravitational/teleport/blob/2cade8a495b6723e6ea791738b6d1eebc6d4d3f3/lib/auth/auth_with_roles.go#L2727)):

```go
// GetRole returns role by name
func (a *ServerWithRoles) GetRole(ctx context.Context, name string) (types.Role, error) {
	// Current-user exception: we always allow users to read roles
	// that they hold.  This requirement is checked first to avoid
	// misleading denial messages in the logs.
	if !apiutils.SliceContainsStr(a.context.User.GetRoles(), name) {
		if err := a.action(apidefaults.Namespace, types.KindRole, types.VerbRead); err != nil {
			return nil, trace.Wrap(err)
		}
	}
	return a.authServer.GetRole(ctx, name)
}
```

We don't treat the names of roles as secret. The user would also only get the roles they got from mapping in `trusted_cluster` resource, which are explicitly given and likely very few.

If we were to do any RBAC checks here, I'm not sure what would they be?

Reply:
The alternative approach would be to move  the db user list logic from `tsh` flow to a backend endpoint. But not sure if we should do it. 

Since roles name and trails are public available for a user logged into root cluster not sure what is the reason to hide `leaf` cluster roles and traits from a user. Thought I think that it would be good idea for now to trim userV2 content and return only to trails and roles name in case of `GetCurrentUser` preventing the case where userV2 will be extended with some sensitive data.


@Tener
Have you check if the [GetUser](https://github.com/gravitational/teleport/blob/master/lib/auth/auth_with_roles.go#L1733) can be leverage here instead of introducing new GetCurrentUser endpoint. At first glace it looks that `authorizeRemoteUser` set the correct  remote user identity: 
https://github.com/gravitational/teleport/blob/2292be1f08ace84bdf6d6ea510e8f554fb7d1e92/lib/auth/permissions.go#L269
https://github.com/gravitational/teleport/blob/2292be1f08ace84bdf6d6ea510e8f554fb7d1e92/lib/auth/permissions.go#L292

where GetUser already allows get get user's own identity
```go
func (a *ServerWithRoles) GetUser(name string, withSecrets bool) (types.User, error) {
      ...
		// if secrets are not being accessed, let users always read
		// their own info.
		if err := a.currentUserAction(name); err != nil {
			// not current user, perform normal permission check.
			if err := a.action(apidefaults.Namespace, types.KindUser, types.VerbRead); err != nil {
				return nil, trace.Wrap(err)
			}
		}
	}
	return a.authServer.Identity.GetUser(name, withSecrets)
```

Reply:
> Have you checked if the [GetUser](https://github.com/gravitational/teleport/blob/master/lib/auth/auth_with_roles.go#L1733) can be leveraged here instead of introducing a new GetCurrentUser endpoint.

Yeah, that was the first thing I tried. `GetUser` and `GetUsers` work against real users, provisioned in one way or another. 

So `a.authServer.Identity.GetUser(name, withSecrets)` goes to backend, which is different from `a.context.User` or `a.context.Identity`.

Reply:
@reedloden @russjones Any objections to exposing the user's own leaf cluster identity in this way, provided we add RBAC checks for the role names? It's needed to fetch leaf's role names so tsh can show "allowed database users" for leaf cluster's databases (needed by Teleport Connect also).

Reply:
> @reedloden @russjones Any objections to exposing the user's own leaf cluster identity in this way, provided we add RBAC checks for the role names? It's needed to fetch leaf's role names so tsh can show "allowed database users" for leaf cluster's databases (needed by Teleport Connect also).

@reedloden @russjones can you take a look as per @r0mant request? It would be good to fix this issue as currently the feature is degraded for remote clusters.

Reply:
@reedloden @russjones PTAL?

Reply:
@Tener I talked with @russjones offline about this and there's no objections to exposing user's remote identities to themselves as long as we require read permissions on roles (basically, what I suggested originally). Can you apply the changes? Then we can merge.

Reply:
> @Tener I talked with @russjones offline about this and there's no objections to exposing user's remote identities to themselves as long as we require read permissions on roles (basically, what I suggested originally). Can you apply the changes? Then we can merge.

Thanks for checking, I'll make the change.